# Project Foundations for Data Science: FoodHub Data Analysis

### Context
The number of restaurants in New York is increasing day by day. Lots of students and busy professionals rely on those restaurants due to their hectic lifestyles. Online food delivery service is a great option for them. It provides them with good food from their favorite restaurants. A food aggregator company FoodHub offers access to multiple restaurants through a single smartphone app.
The app allows the restaurants to receive a direct online order from a customer. The app assigns a delivery person from the company to pick up the order after it is confirmed by the restaurant. The delivery person then uses the map to reach the restaurant and waits for the food package. Once the food package is handed over to the delivery person, he/she confirms the pick-up in the app and travels to the customer's location to deliver the food. The delivery person confirms the drop-off in the app after delivering the food package to the customer. The customer can rate the order in the app. The food aggregator earns money by collecting a fixed margin of the delivery order from the restaurants.

### Objective
The food aggregator company has stored the data of the different orders made by the registered customers in their online portal. They want to analyze the data to get a fair idea about the demand of different restaurants which will help them in enhancing their customer experience. Suppose you are hired as a Data Scientist in this company and the Data Science team has shared some of the key questions that need to be answered. Perform the data analysis to find answers to these questions that will help the company to improve the business.

### Data Description
The data contains the different data related to a food order. The detailed data dictionary is given below.

### Data Dictionary

order_id: Unique ID of the order
customer_id: ID of the customer who ordered the food
restaurant_name: Name of the restaurant
cuisine_type: Cuisine ordered by the customer
cost: Cost of the order
day_of_the_week: Indicates whether the order is placed on a weekday or weekend (The weekday is from Monday to Friday and the weekend is Saturday and Sunday)
rating: Rating given by the customer out of 5
food_preparation_time: Time (in minutes) taken by the restaurant to prepare the food. This is calculated by taking the difference between the timestamps of the restaurant's order confirmation and the delivery person's pick-up confirmation.
delivery_time: Time (in minutes) taken by the delivery person to deliver the food package. This is calculated by taking the difference between the timestamps of the delivery person's pick-up confirmation and drop-off information

### Let us start by importing the required libraries

In [ ]:
# import libraries for data manipulation
import numpy as np
import pandas as pd

# import libraries for data visualization
import matplotlib.pyplot as plt
import seaborn as sns

### Understanding the structure of the data

In [ ]:
# uncomment and run the following lines for Google Colab
#from google.colab import drive
#drive.mount('/content/drive')

In [ ]:
# read the data
df = pd.read_csv(r'C:\Users\Administrator\Desktop\python projects\my first project\foodhub_order.csv')
# returns the first 5 rows
df.head()

#### Observations:
The DataFrame has 9 columns as mentioned in the Data Dictionary. Data in each row corresponds to the order placed by a customer.

### Question 1: How many rows and columns are present in the data? [0.5 mark]

In [ ]:
# Write your code here

num_rows, num_cols = df.shape

print("Number of rows:", num_rows)
print("Number of columns:", num_cols)

Number of rows: 1898
Number of columns: 9


#### Observations:
Number of rows: 1898
Number of columns: 9

### Question 2: What are the datatypes of the different columns in the dataset? (The info() function can be used) [0.5 mark]

In [ ]:
# Use info() to print a concise summary of the DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1898 entries, 0 to 1897
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               1898 non-null   int64  
 1   customer_id            1898 non-null   int64  
 2   restaurant_name        1898 non-null   object 
 3   cuisine_type           1898 non-null   object 
 4   cost_of_the_order      1898 non-null   float64
 5   day_of_the_week        1898 non-null   object 
 6   rating                 1898 non-null   object 
 7   food_preparation_time  1898 non-null   int64  
 8   delivery_time          1898 non-null   int64  
dtypes: float64(1), int64(4), object(4)
memory usage: 133.6+ KB


#### Observations:
dtypes: float64(1), int64(4), object(4)

### Question 3: Are there any missing values in the data? If yes, treat them using an appropriate method. [1 mark]

In [ ]:
# Write your code here
missing_values = df.isnull().sum()
print(missing_values)

order_id                 0
customer_id              0
restaurant_name          0
cuisine_type             0
cost_of_the_order        0
day_of_the_week          0
rating                   0
food_preparation_time    0
delivery_time            0
dtype: int64


#### Observations:
There are no null values in the data set

### Question 4: Check the statistical summary of the data. What is the minimum, average, and maximum time it takes for food to be prepared once an order is placed? [2 marks]

In [ ]:
# Write your code here
summary = df['food_preparation_time'].describe()
print(summary)

count    1898.000000
mean       27.371970
std         4.632481
min        20.000000
25%        23.000000
50%        27.000000
75%        31.000000
max        35.000000
Name: food_preparation_time, dtype: float64


#### Observations:
The minimum, average and maximum time required for food to be prepared once an order is placed are:
20 mins, 27mins and 35mins respectively
count    1898.000000
mean       27.371970
std         4.632481
min        20.000000
25%        23.000000
50%        27.000000
75%        31.000000
max        35.000000
Name: food_preparation_time, dtype: float64

### Question 5: How many orders are not rated? [1 mark]

In [ ]:
# Write the code here
not_rated_orders = (df['rating'] == 'Not given').sum()

print("Number of orders not rated:", not_rated_orders)

Number of orders not rated: 736


#### Observations:
Number of orders not rated: 736

### Exploratory Data Analysis (EDA)

### Univariate Analysis

### Question 6: Explore all the variables and provide observations on their distributions. (Generally, histograms, boxplots, countplots, etc. are used for univariate exploration.) [5 marks]

In [ ]:
# Write th# order_id
# customer_id
plt.figure(figsize=(10, 6))
sns.countplot(data=df, x='customer_id')
plt.xlabel('Customer ID')
plt.ylabel('Frequency')
plt.title('Distribution of Customer IDs')
plt.xticks(rotation=90)
plt.show()

In [ ]:
# restaurant_name
plt.figure(figsize=(12, 6))
sns.countplot(data=df, x='restaurant_name')
plt.xlabel('Restaurant Name')
plt.ylabel('Frequency')
plt.title('Distribution of Restaurants')
plt.xticks(rotation=90)
plt.show()

In [ ]:
# restaurant_name
plt.figure(figsize=(12, 6))
sns.countplot(data=df, x='restaurant_name')
plt.xlabel('Restaurant Name')
plt.ylabel('Frequency')
plt.title('Distribution of Restaurants')
plt.xticks(rotation=90)
plt.show()

In [ ]:
# cuisine_type
plt.figure(figsize=(10, 6))
sns.countplot(data=df, x='cuisine_type')
plt.xlabel('Cuisine Type')
plt.ylabel('Frequency')
plt.title('Distribution of Cuisine Types')
plt.xticks(rotation=90)
plt.show()

In [ ]:
# cost_of_the_order
plt.figure(figsize=(10, 6))
sns.histplot(data=df, x='cost_of_the_order', bins=10)
plt.xlabel('Order Cost')
plt.ylabel('Frequency')
plt.title('Distribution of Order Costs')
plt.show()

In [ ]:
# day_of_the_week
plt.figure(figsize=(10, 6))
sns.countplot(data=df, x='day_of_the_week')
plt.xlabel('Day of the Week')
plt.ylabel('Frequency')
plt.title('Distribution of Orders by Day')
plt.show()

In [ ]:
# rating
plt.figure(figsize=(10, 6))
sns.countplot(data=df, x='rating')
plt.xlabel('Rating')
plt.ylabel('Frequency')
plt.title('Distribution of Ratings')
plt.show()

In [ ]:
# food_preparation_time
plt.figure(figsize=(10, 6))
sns.histplot(data=df, x='food_preparation_time', bins=10)
plt.xlabel('Food Preparation Time')
plt.ylabel('Frequency')
plt.title('Distribution of Food Preparation Time')
plt.show()

In [ ]:
# delivery_time
plt.figure(figsize=(10, 6))
sns.histplot(data=df, x='delivery_time', bins=10)
plt.xlabel('Delivery Time')
plt.ylabel('Frequency')
plt.title('Distribution of Delivery Time')
plt.show()

### Question 7: Which are the top 5 restaurants in terms of the number of orders received? [1 mark]

In [ ]:
# Write the code here
top_restaurants = df['restaurant_name'].value_counts().head(5)
print(top_restaurants)

Shake Shack                  219
The Meatball Shop            132
Blue Ribbon Sushi            119
Blue Ribbon Fried Chicken     96
Parm                          68
Name: restaurant_name, dtype: int64


#### Observations:
The top 5 restaurants in terms of number of orders received are:
Shake Shack                  219
The Meatball Shop            132
Blue Ribbon Sushi            119
Blue Ribbon Fried Chicken     96
Parm                          68

### Question 8: Which is the most popular cuisine on weekends? [1 mark]

In [ ]:
# Write the code here
weekend_cuisine = df[df['day_of_the_week'] == 'Weekend']['cuisine_type']
popular_cuisine_weekend = weekend_cuisine.value_counts().idxmax()

print("The most popular cuisine on weekends is:", popular_cuisine_weekend)

The most popular cuisine on weekends is: American


#### Observations:
The most popular cuisine on weekends is: American

### Question 9: What percentage of the orders cost more than 20 dollars? [1 marks]

In [ ]:
# Write the code here
cost_above_20 = (df['cost_of_the_order'] > 20).mean() * 100

print("Percentage of orders costing more than $20:", cost_above_20, "%")

Percentage of orders costing more than $20: 29.24130663856691 %


#### Observations:
Percentage of orders costing more than $20: 29.24%

### Question 10: What is the mean order delivery time? [1 mark]

In [ ]:
# Write the code here
mean_delivery_time = df['delivery_time'].mean()

print("Mean order delivery time:", mean_delivery_time)

Mean order delivery time: 24.161749209694417


#### Observations:
Mean order delivery time: 24.16

### Question 11: The company has decided to give 20% discount vouchers to the top 3 most frequent customers. Find the IDs of these customers and the number of orders they placed. [1 mark]

In [ ]:
# Write the code here
top_customers = df['customer_id'].value_counts().head(3)
print(top_customers)

52832    13
47440    10
83287     9
Name: customer_id, dtype: int64


#### Observations:
The top 3 most frequent customer id with the number of orders they palced are:
52832    13
47440    10
83287     9

### Multivariate Analysis

### Question 12: Perform a multivariate analysis to explore relationships between the important variables in the dataset. (It is a good idea to explore relations between numerical variables as well as relations between numerical and categorical variables) [5 marks]

In [ ]:
# Write the code here
plt.scatter(df['cost_of_the_order'], df['delivery_time'])
plt.xlabel('Cost of the Order')
plt.ylabel('Delivery Time')
plt.title('Relationship between Cost of the Order and Delivery Time')
plt.show()

In [ ]:
sns.boxplot(x=df['cuisine_type'], y=df['food_preparation_time'])
plt.xlabel('Cuisine Type')
plt.ylabel('Food Preparation Time')
plt.title('Distribution of Food Preparation Time by Cuisine Type')
plt.xticks(rotation=45)
plt.show()

In [ ]:
sns.barplot(x=df['day_of_the_week'], y=df['delivery_time'])
plt.xlabel('Day of the Week')
plt.ylabel('Average Delivery Time')
plt.title('Average Delivery Time by Day of the Week')
plt.show()

In [ ]:
correlation_matrix = df[['cost_of_the_order', 'food_preparation_time', 'delivery_time']].corr()
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm')
plt.title('Correlation Matrix')
plt.show()

In [ ]:
columns_to_compare = ['cost_of_the_order', 'food_preparation_time', 'delivery_time', 'rating']

# Replace "Not given" with NaN in the "rating" column
df['rating'] = df['rating'].replace('Not given', np.nan)

# Convert the "rating" column to numeric
df['rating'] = pd.to_numeric(df['rating'])

# Create a pair plot
sns.pairplot(df[columns_to_compare].dropna())
plt.show()

In [ ]:
#To compare the distribution of ratings across different cuisine types:
sns.countplot(x='cuisine_type', hue='rating', data=df)
plt.xlabel('Cuisine Type')
plt.ylabel('Count')
plt.title('Distribution of Ratings across Cuisine Types')
plt.xticks(rotation=45)
plt.legend(title='Rating')
plt.show()

In [ ]:
#To compare the distribution of ratings on weekdays and weekends:
sns.countplot(x='day_of_the_week', hue='rating', data=df)
plt.xlabel('Days of the week')
plt.ylabel('Count')
plt.title('Distribution of Ratings across Days of the week')
plt.xticks(rotation=45)
plt.legend(title='Rating')
plt.show()

In [ ]:
#To visualize the average rating for each restaurant:
# Replace "Not given" with NaN
df['rating'] = df['rating'].replace('Not given', np.nan)

# Convert the "rating" column to numeric
df['rating'] = pd.to_numeric(df['rating'])

# Group by cuisine type and calculate the average rating
avg_rating_by_cuisine = df.groupby('cuisine_type')['rating'].mean().reset_index()

# Plot the average rating for each cuisine type
plt.figure(figsize=(10, 6))
sns.barplot(x='cuisine_type', y='rating', data=avg_rating_by_cuisine)
plt.xlabel('Cuisine Type')
plt.ylabel('Average Rating')
plt.title('Average Rating by Restaurant')
plt.xticks(rotation=45)
plt.show()

In [ ]:
# to visualize rating on delivery time
subset_df = df[['delivery_time', 'rating']]

# Remove rows with 'Not given' in the rating column
subset_df = subset_df[subset_df['rating'] != 'Not given']

# Convert the rating column to numeric
subset_df['rating'] = pd.to_numeric(subset_df['rating'])

# Create a scatter plot
sns.scatterplot(data=subset_df, x='delivery_time', y='rating')

# Set the plot title and labels
plt.title('Delivery Time vs. Rating')
plt.xlabel('Delivery Time')
plt.ylabel('Rating')

# Display the plot
plt.show()

### Question 13: The company wants to provide a promotional offer in the advertisement of the restaurants. The condition to get the offer is that the restaurants must have a rating count of more than 50 and the average rating should be greater than 4. Find the restaurants fulfilling the criteria to get the promotional offer. [1.5 marks]

In [ ]:
# Write the code here
selected_restaurants = df[df['rating'] != 'Not given'].groupby('restaurant_name').filter(lambda x: len(x) >= 50 and x['rating'].astype(float).mean() > 4)['restaurant_name'].unique()
print(selected_restaurants)

['Blue Ribbon Fried Chicken' 'Shake Shack' 'Blue Ribbon Sushi'
 'The Meatball Shop']


#### Observations:
The restaurants that have rating count more than 50 and average rating greater then 4 are:
'Blue Ribbon Fried Chicken' 
'The Meatball Shop' 
'Shake Shack'
'Blue Ribbon Sushi'

### Question 14: The company charges the restaurant 25% on the orders having cost greater than 20 dollars and 15% on the orders having cost greater than 5 dollars. Find the net revenue generated by the company across all orders. [1.5 marks]

In [ ]:
# Write the code here
# Calculate the commission for orders with cost > 20 dollars (25% commission)
commission_20 = df[df['cost_of_the_order'] > 20]['cost_of_the_order'] * 0.25

# Calculate the commission for orders with cost > 5 dollars (15% commission)
commission_5 = df[(df['cost_of_the_order'] > 5) & (df['cost_of_the_order'] <= 20)]['cost_of_the_order'] * 0.15

# Calculate the net revenue by summing the commissions
net_revenue = commission_20.sum() + commission_5.sum()

print("Net Revenue: $", net_revenue)

Net Revenue: $ 6166.303


#### Observations:
The revenue generated by the company across all the orders is: $ 6166.303

### Question 15: The company wants to analyze the total time required to deliver the food. What percentage of orders take more than 60 minutes to get delivered from the time the order is placed? (The food has to be prepared and then delivered.) [1 marks]

In [ ]:
# Write the code here
# Calculate the total time required for each order (food preparation time + delivery time)
df['total_delivery_time'] = df['food_preparation_time'] + df['delivery_time']

# Count the number of orders that take more than 60 minutes for delivery
num_orders_more_than_60 = (df['total_delivery_time'] > 60).sum()

# Calculate the percentage of orders that take more than 60 minutes for delivery
percentage_orders_more_than_60 = (num_orders_more_than_60 / len(df)) * 100

print("Percentage of orders taking more than 60 minutes for delivery: {:.2f}%".format(percentage_orders_more_than_60))

Percentage of orders taking more than 60 minutes for delivery: 10.54%


#### Observations:
10.54% of orders takes more than 60 mins to get delivered from the time theorder is placed.

### Question 16: The company wants to analyze the delivery time of the orders on weekdays and weekends. How does the mean delivery time vary during weekdays and weekends? [1 marks]

In [ ]:
# Write the code here
# Create separate dataframes for weekdays and weekends
weekday_df = df[df['day_of_the_week'] == 'Weekday']
weekend_df = df[df['day_of_the_week'] == 'Weekend']

# Calculate the mean delivery time for weekdays and weekends
mean_delivery_time_weekday = weekday_df['delivery_time'].mean()
mean_delivery_time_weekend = weekend_df['delivery_time'].mean()

print("Mean delivery time on weekdays: {:.2f} minutes".format(mean_delivery_time_weekday))
print("Mean delivery time on weekends: {:.2f} minutes".format(mean_delivery_time_weekend))

#Calculating time variation between weekdays and weekends:
time_diff = mean_delivery_time_weekday - mean_delivery_time_weekend
print("The time variation during weekdays and weekend is: {:.2f} minutes".format(time_diff))

Mean delivery time on weekdays: 28.34 minutes
Mean delivery time on weekends: 22.47 minutes
The time variation during weekdays and weekend is: 5.87 minutes


#### Observations:
Mean delivery time on weekdays: 28.34 minutes
Mean delivery time on weekends: 22.47 minutes
The delivery time of weekdays is 5.87 more than weekends.

### Conclusion and Recommendations

### Question 17: What are your conclusions from the analysis? What recommendations would you like to share to help improve the business? (You can use cuisine type and feedback ratings to drive your business recommendations.) [5 marks]

### Conclusions:
Based on the analysis, here are some conclusions: 
RESTAURANT ANALYSIS:
1.'Blue Ribbon Fried Chicken' 
 'The Meatball Shop' 
 'Shake Shack'
 'Blue Ribbon Sushi' 
 these are the restaurants which maintains consistant high rating.
 CUSINE TYPE ANALYSIS:

American is the most populae cusine among customers followed by japanese, italian and mexican.
DAYS OF THE WEEK ANALYSIS:
weekends comparitively gets more number of orders. 
RATING ANALYSIS:
rating mostly falls between 3 ans 5 which is good.
CUSTOMER ANALYSIS:
52832, 47440  ,83287 are the customer ids who place frequent orders.
ORDER OF COST ANALYSIS:
The highest frequency of the cost of orders falls between $8 to 18

### Recommendations:
Based on the analysis, here are some recommendations to improve the business:
CUSINE TYPE ANALYSIS:

The most popular cuisine types based on the dataset are Mexican, American, and Japanese.
Focus on these popular cuisine types and consider offering a diverse range of dishes within each cuisine category.
RESTAURANT ANALYSIS:
restaurants with consistently high ratings consider promoting or highlighting them as top-rated restaurants.
DAYS OF THE WEEK ANALYSIS:
weekends comparitively gets more number of orders. so consider promoting and providing offers and discounts on weekends.
RATING ANALYSIS:
rating mostly falls between 3 ans 5 which is good. encourage the customers to rate the orders.
CUSTOMER ANALYSIS:
52832, 47440  ,83287 are the customer ids who place frequent orders. consider offering coupons or discounts to encourag the customers who are consistantly placing orders
ORDER OF COST ANALYSIS:
The highest frequency of the cost of orders falls between $8 to 18. consider adding diverse range of dishes within cost.
FOOD PREPARATION AND DELIVERY TIME ANALYSIS:
optimize food preparation and delivery time as much as possible